src/libopensc/card-cardos5.c

```diff
@@ -0,0 +1,1746 @@
+
+	if (buf->size - buf->bytes_used < *tag_length)
+		return -1;
+
+	if (tag_content != NULL) {
+		if ((*tag_content = malloc(*tag_length)) == NULL)
+			return -1;
+		memcpy(*tag_content, buf->ptr, *tag_length);
+		buf->ptr += *tag_length;
+		buf->bytes_used += *tag_length;
+	}
+
+	return 0;
+}
+
+static int
+cardos5_match_card(sc_card_t *card)
+{
+	if (_sc_match_atr(card, cardos5_atrs, &card->type) < 0)
+		return 0;
+
+	return 1;
+}
+
+static int
+cardos5_init(sc_card_t *card)
+{
+	struct cardos5_private_data	*priv;
+	unsigned int			 flags;
+
+	priv = calloc(1, sizeof(*priv));
+	if (priv == NULL) {
+		sc_log(card->ctx, "calloc");
+		return SC_ERROR_OUT_OF_MEMORY;
+	}
+
+	priv->cse_algorithm = -1U;
+	card->drv_data = priv;
+
+	flags = SC_ALGORITHM_RSA_RAW | SC_ALGORITHM_RSA_HASH_NONE |
+	    SC_ALGORITHM_ONBOARD_KEY_GEN;
+
+	card->name = "CardOS M5";
+	card->caps |= SC_CARD_CAP_APDU_EXT;
+	card->cla = 0x00;
+
+	_sc_card_add_rsa_alg(card,  512, flags, 0);
+	_sc_card_add_rsa_alg(card,  768, flags, 0);
+	_sc_card_add_rsa_alg(card, 1024, flags, 0);
+	_sc_card_add_rsa_alg(card, 1280, flags, 0);
+	_sc_card_add_rsa_alg(card, 1536, flags, 0);
+	_sc_card_add_rsa_alg(card, 1792, flags, 0);
+	_sc_card_add_rsa_alg(card, 2048, flags, 0);
+	_sc_card_add_rsa_alg(card, 2304, flags, 0);
+	_sc_card_add_rsa_alg(card, 2560, flags, 0);
+	_sc_card_add_rsa_alg(card, 2816, flags, 0);
+	_sc_card_add_rsa_alg(card, 3072, flags, 0);
+	_sc_card_add_rsa_alg(card, 3328, flags, 0);
+	_sc_card_add_rsa_alg(card, 3584, flags, 0);
+	_sc_card_add_rsa_alg(card, 3840, flags, 0);
+	_sc_card_add_rsa_alg(card, 4096, flags, 0);
+
+	flags = SC_ALGORITHM_ECDSA_RAW | SC_ALGORITHM_ONBOARD_KEY_GEN;
+
+	_sc_card_add_ec_alg(card, 192, flags, 0);
+	_sc_card_add_ec_alg(card, 224, flags, 0);
+	_sc_card_add_ec_alg(card, 256, flags, 0);
+	_sc_card_add_ec_alg(card, 384, flags, 0);
+	_sc_card_add_ec_alg(card, 512, flags, 0);
+
+	return 0;
+}
+
+static int
+cardos5_finish(sc_card_t *card)
+{
+	free(card->drv_data);
+	card->drv_data = NULL;
+
+	return SC_SUCCESS;
+}
+
+static int
+cardos5_read_record(sc_card_t *card, unsigned int record_number, uint8_t *buf,
+    size_t buf_len, unsigned long flags)
+{
+	struct sc_context	*ctx = card->ctx;
+	sc_apdu_t		 apdu;
+	int			 r;
+
+	memset(&apdu, 0, sizeof(apdu));
+	apdu.cse = SC_APDU_CASE_2_SHORT;
+	apdu.cla = CARDOS5_READ_RECORD_CLA;
+	apdu.ins = CARDOS5_READ_RECORD_INS;
+	apdu.p1 = record_number;
+	apdu.resp = buf;
+	apdu.resplen = buf_len;
+	apdu.le = buf_len;
+
+	if (flags & SC_RECORD_BY_REC_NR)
+		apdu.p2 = 0x04;
+	else if (flags & SC_RECORD_NEXT)
+		apdu.p2 = 0x02;
+
+	if ((r = sc_transmit_apdu(card, &apdu)) != SC_SUCCESS) {
+		sc_log(ctx, "tx/rx error");
+		return r;
+	}
+
+	if ((r = sc_check_sw(card, apdu.sw1, apdu.sw2)) != SC_SUCCESS) {
+		sc_log(ctx, "command failed");
+		return r;
+	}
+
+	return (int)apdu.resplen;
+}
+
+static int
+cardos5_read_binary(sc_card_t *card, unsigned int idx, uint8_t *buf,
+    size_t len, unsigned long flags)
+{
```

Comment:
That's a good point. I suppose sc_read_binary() with card->max_recv_size
= 768 will accomplish exactly what cardos5_read_binary() does. I will
give it a try.